File: pkg/dual_endpoint/6verify.go
package chain_utils

import (
	"fmt"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	ecdsa "github.com/bsv-blockchain/go-sdk/primitives/ecdsa"
	script "github.com/bsv-blockchain/go-sdk/script"
	tx "github.com/bsv-blockchain/go-sdk/transaction"
	sighash "github.com/bsv-blockchain/go-sdk/transaction/sighash"

	multisig "github.com/spycat55/KeymasterMultisigPool/pkg/libs"
)

// verifySignatureWithContext 用于在指定锁定脚本和金额的上下文中验证 DER+SigHash 签名。
// 调用完成后会恢复输入的 SourceTxOutput，确保交易对象不会被持久修改。
func verifySignatureWithContext(
	transactionObject *tx.Transaction,
	inputIndex uint32,
	lockingScript *script.Script,
	sourceSatoshis uint64,
	pub *ec.PublicKey,
	signBytes *[]byte,
) (bool, error) {
	if transactionObject == nil || len(transactionObject.Inputs) == 0 {
		return false, fmt.Errorf("empty transaction or inputs")
	}

	if signBytes == nil || len(*signBytes) < 10 {
		return false, fmt.Errorf("invalid signature length")
	}

	// Expected flag
	flag := sighash.Flag(sighash.ForkID | sighash.All)
	if (*signBytes)[len(*signBytes)-1] != byte(flag) {
		return false, fmt.Errorf("unexpected sighash flag")
	}

	// Backup existing context
	input := transactionObject.Inputs[inputIndex]
	prev := input.SourceTxOutput()

	// Set context needed for sighash
	input.SetSourceTxOutput(&tx.TransactionOutput{Satoshis: sourceSatoshis, LockingScript: lockingScript})

	// Compute sighash and verify DER
	hash, err := transactionObject.CalcInputSignatureHash(inputIndex, flag)
	if err != nil {
		// Restore
		input.SetSourceTxOutput(prev)
		return false, fmt.Errorf("calc sighash failed: %w", err)
	}

	sigDER := (*signBytes)[:len(*signBytes)-1]
	sig, err := ec.ParseDERSignature(sigDER)
	if err != nil {
		input.SetSourceTxOutput(prev)
		return false, fmt.Errorf("parse der failed: %w", err)
	}

	ok := ecdsa.Verify(hash, sig, pub.ToECDSA())

	// Restore original source output
	input.SetSourceTxOutput(prev)
	if !ok {
		return false, fmt.Errorf("signature verify failed")
	}
	return true, nil
}

// ServerVerifyClientSpendSig 用于在 B-Tx 上验证客户端签名。
func ServerVerifyClientSpendSig(
	transactionObject *tx.Transaction,
	totalAmount uint64,
	serverPublicKey *ec.PublicKey,
	clientPublicKey *ec.PublicKey,
	clientSignBytes *[]byte,
) (bool, error) {
	redeem, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return false, fmt.Errorf("build redeem script: %w", err)
	}
	return verifySignatureWithContext(transactionObject, 0, redeem, totalAmount, clientPublicKey, clientSignBytes)
}

// ClientVerifyServerSpendSig 用于在 B-Tx 上验证服务器签名。
func ClientVerifyServerSpendSig(
	transactionObject *tx.Transaction,
	totalAmount uint64,
	serverPublicKey *ec.PublicKey,
	clientPublicKey *ec.PublicKey,
	serverSignBytes *[]byte,
) (bool, error) {
	redeem, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return false, fmt.Errorf("build redeem script: %w", err)
	}
	return verifySignatureWithContext(transactionObject, 0, redeem, totalAmount, serverPublicKey, serverSignBytes)
}

// ServerVerifyClientUpdateSig 用于在更新后的 B-Tx 上验证客户端签名。
func ServerVerifyClientUpdateSig(
	transactionObject *tx.Transaction,
	serverPublicKey *ec.PublicKey,
	clientPublicKey *ec.PublicKey,
	clientSignBytes *[]byte,
) (bool, error) {
	// source satoshis are expected set in input context for updates
	src := transactionObject.Inputs[0].SourceTxOutput()
	if src == nil {
		return false, fmt.Errorf("missing source tx output")
	}
	redeem, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return false, fmt.Errorf("build redeem script: %w", err)
	}
	return verifySignatureWithContext(transactionObject, 0, redeem, src.Satoshis, clientPublicKey, clientSignBytes)
}

// ClientVerifyServerUpdateSig 用于在更新后的 B-Tx 上验证服务器签名。
func ClientVerifyServerUpdateSig(
	transactionObject *tx.Transaction,
	serverPublicKey *ec.PublicKey,
	clientPublicKey *ec.PublicKey,
	serverSignBytes *[]byte,
) (bool, error) {
	src := transactionObject.Inputs[0].SourceTxOutput()
	if src == nil {
		return false, fmt.Errorf("missing source tx output")
	}
	redeem, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return false, fmt.Errorf("build redeem script: %w", err)
	}
	return verifySignatureWithContext(transactionObject, 0, redeem, src.Satoshis, serverPublicKey, serverSignBytes)
}

File: src/dual_endpoint/6verify.ts
import { PublicKey } from '@bsv/sdk/primitives';
import Transaction from '@bsv/sdk/transaction/Transaction';
import Script from '@bsv/sdk/script/Script';
import LockingScript from '@bsv/sdk/script/LockingScript';
import TransactionSignature from '@bsv/sdk/primitives/TransactionSignature';
import { createDualMultisigScript } from './1base_tx';
import { verifyInputSignature } from '../libs/VERIFY';

/**
 * 验证客户端在双端花费交易（B-Tx）上的签名。
 * 需要传入多签 UTXO 的总金额以计算正确的 SigHash。
 */
export function serverVerifyClientSpendSig(
  tx: Transaction,
  totalAmount: number,
  serverPublicKey: PublicKey,
  clientPublicKey: PublicKey,
  clientSignBytes: number[],
): boolean {
  // 使用规范化公钥顺序重新构建多签赎回脚本。
  const redeem: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);

  // 构建临时 SourceTxOutput，上下文会在 verifyInputSignature 内部自行恢复。
  const locking = new LockingScript();
  locking.chunks = redeem.chunks;

  return verifyInputSignature({
    tx,
    inputIndex: 0,
    lockingScript: locking as unknown as Script,
    sourceSatoshis: totalAmount,
    publicKey: clientPublicKey,
    signatureBytes: clientSignBytes,
    expectedSigHashFlag: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });
}

/**
 * 验证服务器在双端花费交易（B-Tx）上的签名。
 */
export function clientVerifyServerSpendSig(
  tx: Transaction,
  totalAmount: number,
  serverPublicKey: PublicKey,
  clientPublicKey: PublicKey,
  serverSignBytes: number[],
): boolean {
  const redeem: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);
  const locking = new LockingScript();
  locking.chunks = redeem.chunks;

  return verifyInputSignature({
    tx,
    inputIndex: 0,
    lockingScript: locking as unknown as Script,
    sourceSatoshis: totalAmount,
    publicKey: serverPublicKey,
    signatureBytes: serverSignBytes,
    expectedSigHashFlag: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });
}

/**
 * 在加载上下文后验证客户端对更新版 B-Tx 的签名。
 * 交易必须已在 inputs[0].sourceTransaction.outputs[0] 中写入金额。
 */
export function serverVerifyClientUpdateSig(
  tx: Transaction,
  serverPublicKey: PublicKey,
  clientPublicKey: PublicKey,
  clientSignBytes: number[],
): boolean {
  const srcValue = tx.inputs[0].sourceTransaction?.outputs[0].satoshis || 0;
  if (!srcValue) return false;
  const redeem: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);
  const locking = new LockingScript();
  locking.chunks = redeem.chunks;

  return verifyInputSignature({
    tx,
    inputIndex: 0,
    lockingScript: locking as unknown as Script,
    sourceSatoshis: srcValue,
    publicKey: clientPublicKey,
    signatureBytes: clientSignBytes,
    expectedSigHashFlag: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });
}

/**
 * 验证服务器对更新版 B-Tx 的签名。
 */
export function clientVerifyServerUpdateSig(
  tx: Transaction,
  serverPublicKey: PublicKey,
  clientPublicKey: PublicKey,
  serverSignBytes: number[],
): boolean {
  const srcValue = tx.inputs[0].sourceTransaction?.outputs[0].satoshis || 0;
  if (!srcValue) return false;
  const redeem: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);
  const locking = new LockingScript();
  locking.chunks = redeem.chunks;

  return verifyInputSignature({
    tx,
    inputIndex: 0,
    lockingScript: locking as unknown as Script,
    sourceSatoshis: srcValue,
    publicKey: serverPublicKey,
    signatureBytes: serverSignBytes,
    expectedSigHashFlag: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });
}
